### ISSUE TITLE
Advices for Memory management in k8s cluster with heavy load of Tekton Pipelines?

### ISSUE THREAD

**Issue by dfuessl** [tektoncd/pipeline#4039]
Hi,
are there any advices for memory management in a k8s cluster with heavy load of Tekton pipelines?

We are running Tekton for analyzing and testing a Java project on certain commits of our git repository.
Each test starts up ~ 30 Oracle databases and many other containers (e. g. running Java). 
When the pipelineRun is finished, the memory usage (seen by kubectl top nodes) does not return to its 
initial state. It increases after each pipelineRun. Eventually we get out of memory kills of processes in our 
containers or pod evictions. Memory usage decreases when we delete old (finished) pipelineRuns 
manually. But we think there could be an automatic management of memory. E. g. when there is not 
enough memory left, the oldest pipelineRuns should be deleted automatically so that there will be no
out of memory kills.

Is this behaviour normal? What are best practices for dealing with these problems?

**Comment by afrittoli** [comment]
Hello @dfuessl - once a `PipelineRun` completes, all the pods will stay around (until the `PipelineRun` is deleted) so that you may access logs, but there should be no residual memory usage though. 

You could check memory usage per pod to make sure:
```
kubectl top pods -n <namespace-where-you-run-pipelines>
```

Tekton pipelines can have a [`finally`](https://tekton.dev/docs/pipelines/pipelines/#adding-finally-to-the-pipeline) section, that you can use to cleanup resources allocated for testing, if that is the cause for the memory leak.

**Comment by dfuessl** [comment]
Hello @afrittoli - thank you very much for your answer, Yes, it is an important feature for us that the pods of a PipelineRun stay around until the Pipeline will be deleted. We often have a look into the logfiles, for example to see what was going on or what went wrong. 

I did do two different things when I tried to analyze residual memory usage on the cluster:

`kubectl top nodes`
`kubectl top pods -A`

I would have expected that by summing up all the values from the second command, I would have a result nearly als much as by summing up all the values from the first command. But in most cases the first sum is much higher than the second sum. At least this is true when some PipelineRuns have been completed and they are still staying around. So I think, there are still some remainders of the completed PipelineRuns that seem to consume some residual memory (at least as seen by the first command).

When the total memory consumption (as seen by `kubectl top nodes`) comes near to 100%, then there will happen OutOfMemory kills or pod evictions. Before that we can sometimes see at least one node showing more than 100% memory usage!

Another observation: 
I connected by a bash session to a node with a high memory value. At this time nothing was being executed on the node (no active pods). With this command
`crictl ps --state Exited`
I saw a really long list of containers. The names showed me that they must have been initSteps, steps and sidecars. Then I deleted all of them (`crictl rmp`). After that the memory usage (as seen by `kubectl top nodes`) for that node dropped significantly and it was nearly as low as I would have expected. So there must be something left in residual memory from completed pipelineRuns (or taskRuns or pods). There might exist two different meanings of "used memory" in the system.

I did this experimental use of `crictl` just out of some kind of curiosity. I think `tkn pr rm` or `kubectl delete pr` will be the right thing to do. I will install a cronjob in k8s which deletes old pipelineRuns. But even then we will always have many completes pipelineRuns around which will "use" a too high amount of memory.

I do not really understand what is going on behind the scenes. I would appreciate if you had any suggestions for me.
